### ISSUE TITLE
no-useless-escape is too aggressive

### ISSUE THREAD

**Issue by feross** [eslint/eslint#7656]
**Tell us about your environment**

* **ESLint Version:** 3.10.2
* **Node Version:** 6.9.1
* **npm Version:** 3.10.8

**What parser (default, Babel-ESLint, etc.) are you using?** Default

**Please show your full configuration:** [`eslint-config-standard` 6.2.1](https://github.com/feross/eslint-config-standard/blob/c879dfb71e03739a45e980687f9a28c94b4a95d2/eslintrc.json)

Specifically:

```json
{
  "no-useless-escape": 2
}
```

**What did you do? Please include the actual source code causing the issue.**

```js
/Expected property "1" of type \[Buffer\], got String "foobar"/
```

**What did you expect to happen?**

No error about a useless escape.

**What actually happened? Please include the actual, raw output from ESLint.**

```
Unnecessary escape character: \]. (no-useless-escape)
```

**Comment by feross** [comment]
The `no-useless-escape` rule got stricter in the latest version of eslint. It's true that it's technically useless to escape `\]` but this is a tad bit aggressive. Let me explain.

ES6 introduced a new "Unicode mode" for regular expressions (which @mathiasbynens recently described as "strict mode" for regex literals). You use it like this `/hello world/u`. Before the recent change in ESLint behavior, `no-useless-escape` was basically equivalent (in my experience) as regexes in "Unicode mode". But now `no-useless-escape` has become very aggressive.

Here's how regexes in "Unicode mode" behave:

`/\a/u` - throws because `a` is **always** an invalid escape sequence. Prevents user from making a mistake.
`/\)/u` - does not throw even though `)` does not need to be escaped. This is because `)` sometimes does need to be escaped and this not likely a user mistake.

Users tend to escape `)` and `]` just to keep their regexes balanced. For example: `/\[Buffer\]/u` looks correct, but `/\[Buffer]/u` does not.

My opinion: if it's good enough for regex "Unicode mode", then it's good enough for ESLint. 👍

**Comment by not-an-aardvark** [comment]
I would say that this is working as intended. The rule is supposed to report useless escape sequences (escape sequences where the backslash can be removed without changing the behavior), and `/\)/` is unambiguously a useless escape sequence.

However, it looks like you're not the only one that considers this behavior to be too aggressive (see https://github.com/nodejs/node/pull/9591). I think making the rule configurable to ignore particular useless escapes (such as `]`) instead of reporting them would be a good enhancement.

**Comment by mickaeltr** [comment]
Hello,

I am having an issue with `no-useless-escape` too, with (I think) a valid Regexp: `/^\/services(\/[^\/]*)?$/'` escaping slashes and considered as a `Unnecessary escape character: \/  no-useless-escape`.

If we're not talking about the same problem, I can indeed open another issue.

Thanks

**Comment by not-an-aardvark** [comment]
@mickaeltr Your `/` is uselessly escaped. `/` does not need to be escaped in a character class, so the corrected version would be `/^\/services(\/[^/]*)?$/`.

**Comment by mickaeltr** [comment]
@not-an-aardvark Great, I didn't know that, thanks for the advice.

**Comment by feross** [comment]
@not-an-aardvark That solution (making the rule configurable to ignore particular useless escapes) would work for me. Then we could configure it to ignore `)` and `]`.

**Comment by silverwind** [comment]
👍  for an option in core eslint to exclude characters. cc: @princejwesley

**Comment by mysticatea** [comment]
I think `no-useless-escape` rule is working perfectly but I will not oppose to add the option ignore specified escapes (maybe `"no-useless-escape": ["error", {"ignore": ["]"]}]`).

FWIW, I don't think `/\[\[(.+?)\]\]/g` is more readable than `/\[\[(.+?)]]/g`.

**Comment by platinumazure** [comment]
Do we need to consider separate options for character classes vs outside character classes?

And I assume the exceptions should only apply to regular expressions and we don't want to allow exclusions for strings- so we should probably ensure the options schema makes that clear.

**Comment by not-an-aardvark** [comment]
> Do we need to consider separate options for character classes vs outside character classes?

I don't think it's necessary to have a separate option for character classes versus non-character classes. If a character has no special meaning regardless of whether it's in a character class, I assume there would be no reason to *only* uselessly-escape it if it's in a character class.

> And I assume the exceptions should only apply to regular expressions and we don't want to allow exclusions for strings- so we should probably ensure the options schema makes that clear.

That seems reasonable, but I wonder if we should have a corresponding option for strings as well (or if we should expand this option to also cover strings).

```js
/* eslint no-useless-escape: ["error", {"ignore": ["]"]}] */

foo("\]"); // no error reported

/\[Buffer\]/; // no error reported
```

**Comment by not-an-aardvark** [comment]
@feross / others who would use this option: Do you have any preference for the options in [my comment](https://github.com/eslint/eslint/issues/7656#issuecomment-263098971) above? In other words, should this option just apply to regular expressions, or should it apply to everything checked by `no-useless-escape`?

**Comment by feross** [comment]
I *think* it should only apply to regular expressions. At least for `standard`, I want to warn if someone writes `var x = '\]\)'` since there's no need to escape in that situation. There's no argument to be made about parallelism.

**Comment by not-an-aardvark** [comment]
While trying to fix https://github.com/eslint/eslint/issues/7789, I took a look at the spec for regular expression parsing. It looks like the ability to parse regular expressions like `/\]/` is actually specified as part of [Annex B](https://tc39.github.io/ecma262/#sec-regular-expressions-patterns) (i.e. it's legacy syntax which is required for browsers, but not required for other JS implementations).

* Regular semantics (required for all JS engines) allow regexes to contain [PatternCharacters](https://tc39.github.io/ecma262/#prod-PatternCharacter), defined as any character other than `^`, `$`, `\`, `.`, `*`, `+`, `?`, `(`, `)`, `[`, `]`, `{`, `}`, `|`.
* Annex B semantics (required for JS engines in browsers, but not all JS engines) allow regex to contain [ExtendedPatternCharacters](https://tc39.github.io/ecma262/#prod-annexB-ExtendedPatternCharacter), defined as any character other than `^`, `$`, `.`, `*`, `+`, `?`, `(`, `)`, `[`, `|`.
* In other words, allowing `]`, `{`, and `}` to be unescaped (outside of regular expressions) is a legacy feature which is not required by the spec. (Note that `{` and `}` are already considered legal escapes by `no-useless-escape`.) Additionally, since unicode-mode regexes don't have the legacy requirement, regular expressions like `/foo]/u` cause a SyntaxError.

So I changed my mind about this -- I think we should allow `]` to be escaped, since it's only "useless" in non-unicode regexes on engines that implement Annex B.

**Comment by ljharb** [comment]
In general, I think eslint rules should only be forbidding things that Annex B makes *not* work - anything that is made optional by Annex B but required absent Annex B must be required (short of very explicit opt-outs by users).

**Comment by not-an-aardvark** [comment]
As far as this particular issue is concerned (we can move the discussion on `]` in particular to https://github.com/eslint/eslint/issues/7789), would anyone use the `exceptions` option for something *other* than `]`? If not, then fixing https://github.com/eslint/eslint/issues/7789 by allowing `]` to be escaped would also make the option unnecessary.

**Comment by not-an-aardvark** [comment]
Now that https://github.com/eslint/eslint/pull/7793 is merged, `no-useless-escape` will no longer warn about escaped `]` in regular expressions.

**Comment by nzakas** [comment]
Can this be closed @not-an-aardvark ?

**Comment by not-an-aardvark** [comment]
I'm not sure. While we relaxed the behavior by default for `]` characters, it's possible someone would still want an option to add exceptions for characters other than `]`.

However, given that I asked a week ago and no one has mentioned that they would want another exception, I suppose it's probably safe to close this.

**Comment by ljharb** [comment]
The other exception I'd want is `)` when not preceded by a `(`, but I don't want it enough to request the issue be reopened :-)

**Comment by feross** [comment]
Yes, seconded. I want `)` but when I tested escaping it earlier like `\)`, it was already being allowed. So I didn't leave a comment. Is this intentional behavior?

**Comment by not-an-aardvark** [comment]
`)` only needs to be escaped outside of character classes.

```js
/)/; // syntax error

/\)/; // ok

/[)]/; // ok

/[\)]/; // useless escape
```

**Comment by feross** [comment]
Got it.

**Comment by abdallahridhamohamed** [comment]
I got the same error:
/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/

**Comment by platinumazure** [comment]
@abdallahridhamohamed Please open a new issue and please make sure to fill out our bug report form as you do so. Thanks!

**Comment by Mouvedia** [comment]
> I will not oppose to add the option ignore

Id need it for `<\/script>`.
